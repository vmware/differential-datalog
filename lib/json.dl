/* JSON parsing/serialization library.
 *
 * Functions in this library are bindings for functions in
 * the Rust `serde_json` crate.
 */

/* Deserialize an instance of type 'T from a string of JSON text.
 *
 * This conversion can fail if the structure of the input does not match the
 * structure expected by 'T, for example if 'T is a struct type but the input
 * contains something other than a JSON map.
 */
extern function from_json_string(json: string): Result<'T, string>

/* Serialize the given data structure as a String of JSON.
 *
 * Serialization can fail if 'T contains a map with non-string keys.
 */
extern function to_json_string(x: 'T): Result<string, string>

/* Represents any valid JSON value.
 */
#[rust="serde(from = \"__json::ValueWrapper\", into = \"__json::ValueWrapper\")"]
typedef JsonValue = // Represents a JSON null value.
                    JsonNull
                  | // Represents a JSON boolean.
                    JsonBool{b: bool}
                  | // Represents a JSON number, whether integer or floating point.
                    JsonNumber{n: JsonNum}
                  | // Represents a JSON string.
                    JsonString{s: istring}
                  | // Represents a JSON array.
                    JsonArray{a: Vec<JsonValue>}
                  | // Represents a JSON object.
                    JsonObject{o: Map<istring, JsonValue>}

/* Get JsonValue value as `bool`.  Returns `None` if value is not `JsonBool`.
 */
function jval_as_bool(v: JsonValue): Option<bool> =
{
    match (v) {
        JsonBool{b} -> Some{b},
        _ -> None
    }
}

/* Get JsonValue value as `bool`.  Returns `def` if value is not `JsonBool`.
 */
function jval_as_bool_or(v: JsonValue, def: bool): bool =
{
    match (v) {
        JsonBool{b} -> b,
        _ -> def
    }
}

/* Get JsonValue value as `JsonNum`.  Returns `None` if value is not `JsonNumber`.
 */
function jval_as_number(v: JsonValue): Option<JsonNum> =
{
    match (v) {
        JsonNumber{n} -> Some{n},
        _ -> None
    }
}

/* Get JsonValue value as `JsonNum`.  Returns `def` if value is not `JsonNumber`.
 */
function jval_as_number_or(v: JsonValue, def: JsonNum): JsonNum =
{
    match (v) {
        JsonNumber{n} -> n,
        _ -> def
    }
}

/* Get JsonValue value as string.  Returns `None` if value is not `JsonString`.
 */
function jval_as_string(v: JsonValue): Option<istring> =
{
    match (v) {
        JsonString{s} -> Some{s},
        _ -> None
    }
}

/* Get JsonValue value as string.  Returns `def` if value is not `JsonString`.
 */
function jval_as_string_or(v: JsonValue, def: istring): istring =
{
    match (v) {
        JsonString{s} -> s,
        _ -> def
    }
}

/* Get JsonValue value as array.  Returns `None` if value is not `JsonArray`.
 */
function jval_as_array(v: JsonValue): Option<Vec<JsonValue>> =
{
    match (v) {
        JsonArray{a} -> Some{a},
        _ -> None
    }
}

/* Get JsonValue value as array.  Returns `def` if value is not `JsonArray`.
 */
function jval_as_array_or(v: JsonValue, def: Vec<JsonValue>): Vec<JsonValue> =
{
    match (v) {
        JsonArray{a} -> a,
        _ -> def
    }
}

/* Get JsonValue value as object.  Returns `None` if value is not `JsonObject`.
 */
function jval_as_object(v: JsonValue): Option<Map<istring, JsonValue>> =
{
    match (v) {
        JsonObject{o} -> Some{o},
        _ -> None
    }
}

/* Get JsonValue value as object.  Returns `def` if value is not `JsonObject`.
 */
function jval_as_object_or(v: JsonValue, def: Map<istring, JsonValue>): Map<istring, JsonValue> =
{
    match (v) {
        JsonObject{o} -> o,
        _ -> def
    }
}

/* Get attribute by name.  Returns `None` if value is not `JsonObject` or
 * does not have the specified attribute.
 */
function jval_get(v: JsonValue, attr: istring): Option<JsonValue> =
{
    match (v) {
        JsonObject{o} -> map_get(o, attr),
        _ -> None
    }
}

/* Get attribute by name.  Returns `def` if value is not `JsonObject` or
 * does not have the specified attribute.
 */
function jval_get_or(v: JsonValue, attr: istring, def: JsonValue): JsonValue =
{
    match (v) {
        JsonObject{o} -> option_unwrap_or(map_get(o, attr), def),
        _ -> def
    }
}

typedef JsonNum = // Integer number.  The value must be in the range between the
                  // smallest `s64` number and the largest `u64` number,
                  // inclusive.  Values outside of this range will be truncated
                  // during serialization.
                  JsonInt{i: s128}
                | // NaN and infinity are not valid JSON values.
                  // When serializing to JSON, such values will be replaced with 0.
                  JsonFloat{d: double}

/* Wrapper type that serializes its inner object into binary formats via JSON string.
 *
 * This is a workaround for a `serde` limitation that prevents certain data types from
 * being serialized or deserialized from non-self-describing data formats like bincode.
 * In particular, this is the case for data types annotated with JSON-related serde
 * attributes like `tag` and `flatten`.  In practice this means that types annotated for
 * serialization to JSON cannot be serialized to other formats.
 *
 * The workaround is to use different serialization logic when serializing to binary and
 * non-binary formats.  In the binary case, we want to serialize such types into a string
 * and then serialize the string as a field in binary data.  In the non-binary case, we
 * stick to normal serde behavior.  `JsonWrapper` implements this choice by checking the
 * `is_human_readable` property of the serializer.
 *
 * The current implementation has an important flaw: serde currently does not provide a
 * reliable way to determine if a serializer supports self-describing data.  We use
 * `is_human_readable` attribute as an approximation, which is not always correct: binary
 * formats can be self-describing (BSON, Pickle), conversely not all human-readable formats
 * are self-describing.  Nevertheless, the current design enables useful scenarios, like
 * transcoding JSON to bincode.
 */
#[custom_serde]
typedef JsonWrapper<'T> = JsonWrapper{x: 'T}
