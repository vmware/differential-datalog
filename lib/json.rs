pub fn json_from_json_string<'de, T: serde::Deserialize<'de>>(
    json: &'de String,
) -> std_Result<T, String> {
    res2std(serde_json::from_str::<'de>(&*json))
}

pub fn json_to_json_string<T: serde::Serialize>(x: &T) -> std_Result<String, String> {
    res2std(serde_json::to_string(x))
}

pub struct ValueWrapper(serde_json::value::Value);

impl serde::Serialize for ValueWrapper {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        if serializer.is_human_readable() {
            self.0.serialize(serializer)
        } else {
            serde_json::to_string(&self.0)
                .map_err(|e| serde::ser::Error::custom(e))?
                .serialize(serializer)
        }
    }
}

impl<'de> Deserialize<'de> for ValueWrapper {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        if deserializer.is_human_readable() {
            Ok(ValueWrapper(serde_json::Value::deserialize(deserializer)?))
        } else {
            Ok(ValueWrapper(
                serde_json::from_str::<serde_json::Value>(
                    String::deserialize(deserializer)?.as_ref(),
                )
                .map_err(|e| serde::de::Error::custom(e))?,
            ))
        }
    }
}

impl From<ValueWrapper> for json_JsonValue {
    fn from(x: ValueWrapper) -> Self {
        json_JsonValue::from(x.0)
    }
}

impl From<json_JsonValue> for ValueWrapper {
    fn from(x: json_JsonValue) -> Self {
        ValueWrapper(serde_json::Value::from(x))
    }
}

impl From<serde_json::value::Value> for json_JsonValue {
    fn from(x: serde_json::value::Value) -> Self {
        match x {
            serde_json::value::Value::Null => json_JsonValue::json_JsonNull,
            serde_json::value::Value::Bool(b) => json_JsonValue::json_JsonBool { b },
            serde_json::value::Value::Number(n) => json_JsonValue::json_JsonNumber {
                n: json_JsonNum::from(n),
            },
            serde_json::value::Value::String(s) => json_JsonValue::json_JsonString {
                s: internment_intern(&s),
            },
            serde_json::value::Value::Array(a) => {
                let v: Vec<json_JsonValue> =
                    a.into_iter().map(|v| json_JsonValue::from(v)).collect();
                json_JsonValue::json_JsonArray {
                    a: std_Vec::from(v),
                }
            }
            serde_json::value::Value::Object(o) => json_JsonValue::json_JsonObject {
                o: o.into_iter()
                    .map(|(k, v)| (internment_intern(&k), json_JsonValue::from(v)))
                    .collect(),
            },
        }
    }
}

impl From<json_JsonValue> for serde_json::value::Value {
    fn from(x: json_JsonValue) -> Self {
        match x {
            json_JsonValue::json_JsonNull => serde_json::value::Value::Null,
            json_JsonValue::json_JsonBool { b } => serde_json::value::Value::Bool(b),
            json_JsonValue::json_JsonNumber { n } => json_val_from_num(n),
            json_JsonValue::json_JsonString { s } => {
                serde_json::value::Value::String(internment_ival(&s).clone())
            }
            json_JsonValue::json_JsonArray { a } => serde_json::value::Value::Array(
                a.into_iter()
                    .map(|v| serde_json::value::Value::from(v))
                    .collect(),
            ),
            json_JsonValue::json_JsonObject { o } => serde_json::value::Value::Object(
                o.into_iter()
                    .map(|(k, v)| {
                        (
                            internment_ival(&k).clone(),
                            serde_json::value::Value::from(v),
                        )
                    })
                    .collect(),
            ),
        }
    }
}

impl From<serde_json::Number> for json_JsonNum {
    fn from(n: serde_json::Number) -> Self {
        if n.is_u64() {
            json_JsonNum::json_JsonInt {
                i: n.as_u64().unwrap() as i128,
            }
        } else if n.is_i64() {
            json_JsonNum::json_JsonInt {
                i: n.as_i64().unwrap() as i128,
            }
        } else if n.is_f64() {
            json_JsonNum::json_JsonFloat {
                d: OrderedFloat::from(n.as_f64().unwrap()),
            }
        } else {
            panic!("JsonNum::from::<Number>(): unknown number format: '{}'", n)
        }
    }
}

fn json_val_from_num(n: json_JsonNum) -> serde_json::value::Value {
    match n {
        json_JsonNum::json_JsonInt { i } => {
            if i >= 0 {
                serde_json::value::Value::Number(serde_json::Number::from(i as u64))
            } else {
                serde_json::value::Value::Number(serde_json::Number::from(i as i64))
            }
        }
        json_JsonNum::json_JsonFloat { d } => serde_json::Number::from_f64(*d)
            .map(serde_json::value::Value::Number)
            .unwrap_or_else(|| serde_json::value::Value::Null),
    }
}

/* Some JSON schemas represent integers as strings.  This module serializes/deserializes
 * integers or any other types that implement `FromStr` and `ToString` traits to/from strings.
 * To use this module, add the following annotation to the DDlog field that needs to be
 * deserialized from a string:
 *
 * ```
 * typedef U64FromString = U64FromString {
 *   x: string,
 *   #[rust="serde(with=\"serde_string\")"]
 *   y: u64
 * }
 * ```
 */
pub mod serde_string {
    use serde::de;
    use serde::Deserialize;
    use serde::Deserializer;
    use serde::Serialize;
    use serde::Serializer;

    pub fn serialize<S, T>(x: &T, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
        T: std::string::ToString,
    {
        x.to_string().serialize(serializer)
    }

    pub fn deserialize<'de, D, T>(deserializer: D) -> Result<T, D::Error>
    where
        D: Deserializer<'de>,
        T: std::str::FromStr,
        <T as std::str::FromStr>::Err: std::fmt::Display,
    {
        T::from_str(&String::deserialize(deserializer)?).map_err(|e| de::Error::custom(e))
    }
}

impl<T: Serialize> Serialize for json_JsonWrapper<T> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        if serializer.is_human_readable() {
            self.x.serialize(serializer)
        } else {
            serde_json::to_string(&self.x)
                .map_err(|e| serde::ser::Error::custom(e))?
                .serialize(serializer)
        }
    }
}

impl<'de, T: DeserializeOwned> Deserialize<'de> for json_JsonWrapper<T> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        if deserializer.is_human_readable() {
            Ok(json_JsonWrapper {
                x: T::deserialize(deserializer)?,
            })
        } else {
            Ok(json_JsonWrapper {
                x: serde_json::from_str::<T>(String::deserialize(deserializer)?.as_ref())
                    .map_err(|e| serde::de::Error::custom(e))?,
            })
        }
    }
}
