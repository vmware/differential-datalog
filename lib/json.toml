[dependencies.serde_json]
version = "1.0"
