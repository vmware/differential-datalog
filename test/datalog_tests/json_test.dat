dump json_test::JsonTest;
dump json_test::JsonTestValue;

start;
insert json_test::Deserialized[@json"{\"@type\": \"t.V1\", \"b\": true}"],
commit;

dump json_test::ODeserialized;
