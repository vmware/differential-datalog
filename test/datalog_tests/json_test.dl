import json

output relation JsonTest(description: string, value: string)

function to_json_string_or_default(x: 'T): string =
    result_unwrap_or_default(to_json_string(x))


typedef BoolStruct = BoolStruct {
    b: bool
}

function boolStruct1(): string = [|{"b":true}|]
function boolStruct2(): string = [|{"b":true, "foo":"bar"}|]
function boolStruct3(): string = [|{"foo":"bar"}|]

JsonTest(boolStruct1(),
         to_json_string_or_default(from_json_string(boolStruct1()): Result<BoolStruct, string>)).
JsonTest(boolStruct2(),
         to_json_string_or_default(from_json_string(boolStruct2()): Result<BoolStruct, string>)).
JsonTest(boolStruct3(),
         to_json_string_or_default(from_json_string(boolStruct3()): Result<BoolStruct, string>)).

typedef Top = Top {
    t: string,
    id: string,
    nested: Nested
}

typedef Nested = Nested {
    x: string,
    z: u64
}

function nestedStruct1(): string = [|{"t":"foo", "id":"1001001001", "nested": {"x": "x", "z": 100000}}|]

JsonTest(nestedStruct1(),
         to_json_string_or_default(from_json_string(nestedStruct1()): Result<Top, string>)).

typedef TopFlat = TopFlat {
    t: string,
    #[rust="serde(rename=\"@id\")"]
    id: string,
    #[rust="serde(flatten)"]
    nested: Nested
}

function flattenedStruct1(): string = [|{"t":"foo", "@id":"1001001001", "x": "x", "z": 100000}|]

JsonTest(flattenedStruct1(),
         to_json_string_or_default(from_json_string(flattenedStruct1()): Result<TopFlat, string>)).

typedef Array = Vec<BoolStruct>

function array1(): string = [|[{"b":true}, {"b":false}, {"b":true}, {"b":false}]|]

JsonTest(array1(),
         to_json_string_or_default(from_json_string(array1()): Result<Array, string>)).

typedef BoolMap = Map<string, BoolStruct>

function map1(): string = [|{"1":{"b":true}, "2":{"b":false}, "3":{"b":true}, "4":{"b":false}}|]

JsonTest(map1(),
         to_json_string_or_default(from_json_string(map1()): Result<BoolMap, string>)).

typedef Enum = #[rust="serde(rename = \"Variant1\")"]
               Variant1 { b: bool }
             | #[rust="serde(rename = \"Variant2\")"]
               Variant2 { u: u32 }

function enum1(): string = [|{"Variant1": {"b": true}}|]
function enum2(): string = [|{"Variant2": {"u": 100}}|]

JsonTest(enum1(),
         to_json_string_or_default(from_json_string(enum1()): Result<Enum, string>)).
JsonTest(enum2(),
         to_json_string_or_default(from_json_string(enum2()): Result<Enum, string>)).

#[rust="serde(tag = \"@type\")"]
typedef TaggedEnum = #[rust="serde(rename = \"t.V1\")"]
                     TVariant1 { b: bool }
                   | #[rust="serde(rename = \"t.V2\")"]
                     TVariant2 { u: u32 }

function tenum1(): string = [|{"@type": "t.V1", "b": true}|]
function tenum2(): string = [|{"@type": "t.V2", "u": 100}|]
function tenum3(): string = [|{"@type": "t.V2", "b": false}|]

JsonTest(tenum1(),
         to_json_string_or_default(from_json_string(tenum1()): Result<TaggedEnum, string>)).
JsonTest(tenum2(),
         to_json_string_or_default(from_json_string(tenum2()): Result<TaggedEnum, string>)).
JsonTest(tenum3(),
         to_json_string_or_default(from_json_string(tenum3()): Result<TaggedEnum, string>)).


function scalar1(): string = [|true|]
function scalar2(): string = [|100|]
function scalar3(): string = [|-100|]
function scalar4(): string = [|"foo bar"|]
function scalar5(): string = [|2.99792458e8|]

JsonTest(scalar1(),
         to_json_string_or_default(from_json_string(scalar1()): Result<JsonValue, string>)).
JsonTest(scalar2(),
         to_json_string_or_default(from_json_string(scalar2()): Result<JsonValue, string>)).
JsonTest(scalar3(),
         to_json_string_or_default(from_json_string(scalar3()): Result<JsonValue, string>)).
JsonTest(scalar4(),
         to_json_string_or_default(from_json_string(scalar4()): Result<JsonValue, string>)).
JsonTest(scalar5(),
         to_json_string_or_default(from_json_string(scalar5()): Result<JsonValue, string>)).

input relation Deserialized[TaggedEnum]
output relation ODeserialized[TaggedEnum]
ODeserialized[x] :- Deserialized[x].

typedef Optional = Optional {
    s: Option<string>,
    i: Option<u64>,
    v: Option<JsonValue>
}

function optional1(): string = [|{}|]
function optional2(): string = [|{"s": "foo"}|]
function optional3(): string = [|{"s": "foo", "i": 100000}|]
function optional4(): string = [|{"s": "foo", "i": 100000, "v": 2.5}|]

JsonTest(optional1(),
         to_json_string_or_default(from_json_string(optional1()): Result<Optional, string>)).
JsonTest(optional2(),
         to_json_string_or_default(from_json_string(optional2()): Result<Optional, string>)).
JsonTest(optional3(),
         to_json_string_or_default(from_json_string(optional3()): Result<Optional, string>)).
JsonTest(optional4(),
         to_json_string_or_default(from_json_string(optional4()): Result<Optional, string>)).

typedef StructWithKey = StructWithKey {
    key: u64,
    payload: string
}

function key_structWithKey(x: StructWithKey): u64 =
{
    x.key
}

typedef StructWithMap = StructWithMap {
    #[deserialize_from_array=key_structWithKey()]
    f: Map<u64, StructWithKey>
}

function struct_with_map1(): string = [|{"f": [{"key": 100, "payload": "foo"}]}|]

JsonTest(struct_with_map1(),
         to_json_string_or_default(from_json_string(struct_with_map1()): Result<StructWithMap, string>)).


typedef U64FromString = U64FromString {
    x: string,
    #[rust="serde(with=\"serde_string\")"]
    y: u64
}

function u64FromString1(): string = [|{"x": "x", "y": "100000"}|]
function u64FromString2(): string = [|{"x": "x", "y": "-100000"}|]

JsonTest(u64FromString1(),
         to_json_string_or_default(from_json_string(u64FromString1()): Result<U64FromString, string>)).
JsonTest(u64FromString2(),
         to_json_string_or_default(from_json_string(u64FromString2()): Result<U64FromString, string>)).

typedef S32FromString = S32FromString {
    #[rust="serde(with=\"serde_string\")"]
    x: s32
}

function s32FromString1(): string = [|{"x": "-100000"}|]
function s32FromString2(): string = [|{"x": "100000000000"}|]

JsonTest(s32FromString1(),
         to_json_string_or_default(from_json_string(s32FromString1()): Result<S32FromString, string>)).
JsonTest(s32FromString2(),
         to_json_string_or_default(from_json_string(s32FromString2()): Result<S32FromString, string>)).

typedef WrappedEnum = JsonWrapper<TaggedEnum>

function wenum1(): string = [|{"@type": "t.V1", "b": true}|]
function wenum2(): string = [|{"@type": "t.V2", "u": 100}|]
function wenum3(): string = [|{"@type": "t.V2", "b": false}|]

JsonTest("wrapped " ++ wenum1(),
         to_json_string_or_default(from_json_string(wenum1()): Result<WrappedEnum, string>)).
JsonTest("wrapped " ++ wenum2(),
         to_json_string_or_default(from_json_string(wenum2()): Result<WrappedEnum, string>)).
JsonTest("wrapped " ++ wenum3(),
         to_json_string_or_default(from_json_string(wenum3()): Result<WrappedEnum, string>)).
