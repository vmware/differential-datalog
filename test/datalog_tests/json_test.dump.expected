json_test.JsonTest{.description = "\"foo bar\"", .value = "{\"std_Ok\":{\"res\":\"foo bar\"}}"}
json_test.JsonTest{.description = "-100", .value = "{\"std_Ok\":{\"res\":-100}}"}
json_test.JsonTest{.description = "100", .value = "{\"std_Ok\":{\"res\":100}}"}
json_test.JsonTest{.description = "2.99792458e8", .value = "{\"std_Ok\":{\"res\":299792458.0}}"}
json_test.JsonTest{.description = "[{\"b\":true}, {\"b\":false}, {\"b\":true}, {\"b\":false}]", .value = "{\"std_Ok\":{\"res\":[{\"b\":true},{\"b\":false},{\"b\":true},{\"b\":false}]}}"}
json_test.JsonTest{.description = "true", .value = "{\"std_Ok\":{\"res\":true}}"}
json_test.JsonTest{.description = "wrapped {\"@type\": \"t.V1\", \"b\": true}", .value = "{\"std_Ok\":{\"res\":{\"@type\":\"t.V1\",\"b\":true}}}"}
json_test.JsonTest{.description = "wrapped {\"@type\": \"t.V2\", \"b\": false}", .value = "{\"std_Err\":{\"err\":\"missing field `u`\"}}"}
json_test.JsonTest{.description = "wrapped {\"@type\": \"t.V2\", \"u\": 100}", .value = "{\"std_Ok\":{\"res\":{\"@type\":\"t.V2\",\"u\":100}}}"}
json_test.JsonTest{.description = "{\"1\":{\"b\":true}, \"2\":{\"b\":false}, \"3\":{\"b\":true}, \"4\":{\"b\":false}}", .value = "{\"std_Ok\":{\"res\":{\"1\":{\"b\":true},\"2\":{\"b\":false},\"3\":{\"b\":true},\"4\":{\"b\":false}}}}"}
json_test.JsonTest{.description = "{\"@type\": \"t.V1\", \"b\": true}", .value = "{\"std_Ok\":{\"res\":{\"@type\":\"t.V1\",\"b\":true}}}"}
json_test.JsonTest{.description = "{\"@type\": \"t.V2\", \"b\": false}", .value = "{\"std_Err\":{\"err\":\"missing field `u`\"}}"}
json_test.JsonTest{.description = "{\"@type\": \"t.V2\", \"u\": 100}", .value = "{\"std_Ok\":{\"res\":{\"@type\":\"t.V2\",\"u\":100}}}"}
json_test.JsonTest{.description = "{\"Variant1\": {\"b\": true}}", .value = "{\"std_Ok\":{\"res\":{\"Variant1\":{\"b\":true}}}}"}
json_test.JsonTest{.description = "{\"Variant2\": {\"u\": 100}}", .value = "{\"std_Ok\":{\"res\":{\"Variant2\":{\"u\":100}}}}"}
json_test.JsonTest{.description = "{\"b\":true, \"foo\":\"bar\"}", .value = "{\"std_Ok\":{\"res\":{\"b\":true}}}"}
json_test.JsonTest{.description = "{\"b\":true}", .value = "{\"std_Ok\":{\"res\":{\"b\":true}}}"}
json_test.JsonTest{.description = "{\"f\": [{\"key\": 100, \"payload\": \"foo\"}]}", .value = "{\"std_Ok\":{\"res\":{\"f\":[{\"key\":100,\"payload\":\"foo\"}]}}}"}
json_test.JsonTest{.description = "{\"foo\":\"bar\"}", .value = "{\"std_Err\":{\"err\":\"missing field `b` at line 1 column 13\"}}"}
json_test.JsonTest{.description = "{\"s\": \"foo\", \"i\": 100000, \"v\": 2.5}", .value = "{\"std_Ok\":{\"res\":{\"s\":\"foo\",\"i\":100000,\"v\":2.5}}}"}
json_test.JsonTest{.description = "{\"s\": \"foo\", \"i\": 100000}", .value = "{\"std_Ok\":{\"res\":{\"s\":\"foo\",\"i\":100000,\"v\":null}}}"}
json_test.JsonTest{.description = "{\"s\": \"foo\"}", .value = "{\"std_Ok\":{\"res\":{\"s\":\"foo\",\"i\":null,\"v\":null}}}"}
json_test.JsonTest{.description = "{\"t\":\"foo\", \"@id\":\"1001001001\", \"x\": \"x\", \"z\": 100000}", .value = "{\"std_Ok\":{\"res\":{\"t\":\"foo\",\"@id\":\"1001001001\",\"x\":\"x\",\"z\":100000}}}"}
json_test.JsonTest{.description = "{\"t\":\"foo\", \"id\":\"1001001001\", \"nested\": {\"x\": \"x\", \"z\": 100000}}", .value = "{\"std_Ok\":{\"res\":{\"t\":\"foo\",\"id\":\"1001001001\",\"nested\":{\"x\":\"x\",\"z\":100000}}}}"}
json_test.JsonTest{.description = "{\"x\": \"-100000\"}", .value = "{\"std_Ok\":{\"res\":{\"x\":\"-100000\"}}}"}
json_test.JsonTest{.description = "{\"x\": \"100000000000\"}", .value = "{\"std_Err\":{\"err\":\"number too large to fit in target type at line 1 column 21\"}}"}
json_test.JsonTest{.description = "{\"x\": \"x\", \"y\": \"-100000\"}", .value = "{\"std_Err\":{\"err\":\"invalid digit found in string at line 1 column 26\"}}"}
json_test.JsonTest{.description = "{\"x\": \"x\", \"y\": \"100000\"}", .value = "{\"std_Ok\":{\"res\":{\"x\":\"x\",\"y\":\"100000\"}}}"}
json_test.JsonTest{.description = "{}", .value = "{\"std_Ok\":{\"res\":{\"s\":null,\"i\":null,\"v\":null}}}"}
json_test.TVariant1{.b = true}
